### PR TITLE
Allows users to change default GUID and UID for the JBoss EAP user

### PR DESCRIPTION
**Pull request by rpelisse**
@sabre1041 let's discuss here. As you can see if I nest the jcliff: module within the jboss_eap role, the use will need to provide the EAP/Wildfly subsystems configuration information, which is going to be complicated to specify as a data structures.

Rather than having jcliff: lives within the tasks/jboss_instances, I think it should be the only tasks defined into the playbook, which will mean something like that:

```
---
- name: "Ansible EAP Role demo"
  hosts: "{{ wildfly_ansible_hosts | default('localhost') }}"
  roles:
    - sabre1041.redhat-csp-download
    - redhat-cop.jcliff
    - redhat-cop.jboss_eap
  tasks:
    - name: "Define instance configuration"
      jcliff:
        wfly_home: "{{ jboss_eap_home }}"
        subsystems:
```

I have a feeling you would like to keep everything hidden within the jboss_eap role, but I don't know how to achieve that!

### CODE REVIEW

**Comment by sabre1041** [comment]
@rpelisse I recommend that our role be fairly generic with the role `dependencies` listed in the `meta/main.yml` file as you have it. We define a new variable (say `jboss_subsystems`) that is either an empty list, or has a minimal configuration that would be applicable. Users can then define the `jboss_subsystems` in their own playbook or through `vars` files that override the defaults.

What do you think?

**Comment by rpelisse** [comment]
Try to implement this jboss_subsystems var and I think you will quickly see what the issue is ;) Imagine that the user only provide a subset of subsystems for instance - how would it work on the Jcliff module ? We'll end up having subsystem declared with null values? That's going to be messy, we'll need to modify the jcliff module to see those invalid or incomplete data structure as "OK"...

Also, if we "hide" the jcliff module deep into the role this way, we lose all the work we did to make it "ansible friendly". We also use the nice "user experience" coming with it. Now the user has to deal with a complex data structure, with no validation whatsoever. Everytime he is going to make a typo, he might get cryptic messages. If he uses directly the jcliff module, he has a clear outline, each subsystem has a matching resources within the module. It's far more nicer than a huge jboss_subsystems structure. 

I do understand the urge to have just a role and "voilà" everything is ready and running, but I feel, in this particular case, it cost us much more than what it bring us.

If we want to do fairly generic, I would suggest if leave the jcliff module out of the jboss_eap configuration. Tweaking subsystems is not generic. Generic is start default server, with default configuration. If the user wants to do more, like deploying an app, then he should add the jcliff module into its playbook. (we could leave the dependency to automate-jcliff for this very purpose, or simply document how to do it).

**Comment by rpelisse** [comment]
@sabre1041 not sure you'll be notified from my comment above so I've added this. Also, if I am basically advocating to leave jcliff out of this repo (to keep it generic), I think it should definitly appears in the [ansible middleware playbooks](https://github.com/redhat-cop/ansible-middleware-playbooks). We could imagine having a playbook such as jboss-eap-7.1-with-datasource-and-custom-properties-on-rhel7.yml, for instance.

**Comment by sabre1041** [comment]
@rpelisse I am on the fence on whether it should be included or not. 

My thought is that it would be nice if it had it included as then end users do not need to add the `jcliff` role to their playbooks or really need to worry about `jcliff` other than defining a variable that would then be executed by the base role (lower the barrier of entry [include role and a variable and you have EAP]. ).

The default value in `jboss_eap` should be empty (basically a noop), but it gives the end user the ability to define the variable within their playbook or `vars` file as described previously.

**Comment by rpelisse** [comment]
@sabre1041 I’m happy that you are on the fence about this, because I’m also on the fence about it. At first, it seemed obvious we should do so, but the more I looked into it, the less it made sense to me. As I’m going to try to sum up below, I think we lose more than we gain by adding automate-jcliff into jboss_eap.

First, let me try to explain what I’m worried about regarding the jboss_eap var approach. Let’s see a user wants to add a couple of system properties using this var, here what it would look-like in his playbook:
```
jboss_eap:
  -  { subsystem: 'system-props', vars: {
    - name: 'FirstVar', value: 'FirstValue'
    - name: 'SecondVar', value: 'SecondValue' }
    -  name: 'ThirdVar', value: 'ThirdValue' }
...
}
```
(There might be better away of structuring the data and this might not be yaml compliant, but that’s beside my point here.)

Now, let’s look at the associated code we’ll have to add to the role:
```
 - name: "Tuning Wildfly using JCliff"
      jcliff:
        wfly_home: "{{ jboss_eap_home }}"
# all the following attributes needs a default value that can be override by user using the jboss_eap var (or some other vars):
        jcliff_home: ...
        jcliff:  ...
       management_username: ...
       management_password: ...
       rules_dir: ...
       management_host: 
       management_port:
       jcliff_jvm:...
       subsystems:
          - drivers: # there is no value in the user jboss_eap vars, how do we deal with that? We could have empty array, it may work, but it will mean we need to make them OK in the jcliff module, which arms the resilience / validation of the module itself....
          - datasources: # same
          - system_props: "{{ jboss_eap.subsystems['system_props'] }}"
...
```
The issues mentioned in the code above is my first concern. Let’s move to the second one.

Assuming we managed to find a syntax that works for the jboss_eap var and the module and we implement it into the jboss_eap role, what do we get out of it?

- An empty jcliff: configuration that does nothing (clutter the code)
- A role dependency added plus some package installed (jcliff) that the user may not use (still installed).
- User needs to use a **data structure** (in opposite of filling the appropriate fields of jcliff) in order to configured its instance, which IMHO is a degradation in terms of user experience. (Plus we need to maintain a sort of documentation for the data structure which adds to interface available with the jcliff module).
- arguably the user gets to have jcliff "already installed" if he needs it.

(As a side note, we also need to constantly tweak this data structure, every time we add a new subsystem or make a change into jcliff).

On top of those concerns, I think it would be a better approach to respect the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) when it comes to the role:

- jboss_eap installs jboss_eap (and deal with patches) - which its current function AFAIU
- automate_jcliff takes a WFLY/EAP instance (most likely installed by previous roles) and allow to fine-tune the instance

If do that, we will avoid cluttering the code of the role by adding more or more responsibility. We’ll get simple, clear-cut independent roles, that can work together.

In short, I think we should rather focus on demonstrating **how to use jboss_eap and automate_jcliff together**, rather than trying to hide the automate_jcliff inside the jboss_eap role. What do you think?

**Comment by sabre1041** [comment]
@rpelisse lets go with the split approach. we can revisit it later if we decide to, but I think that indeed makes the most sense and best approach. Having solid docs and examples with make it even more straightforward for end users.

**Comment by sabre1041** [comment]
@rpelisse do we want to close this issue?

**Comment by rpelisse** [comment]
@sabre1041 There is one (small) changes to jboss_eap, I've proposed along with the jcliff integratiion. So I've removed the commit about jcliff and let only this change (And renamed accordingly this PR).
